### PR TITLE
Improve user share suggestions

### LINKED ISSUE #1963: Improve user share suggestions
1. Entering the email of a local user should bring up the user as a local share
2. Entering the cloud id of a local user should bring up the user as a local share
3. Searching for `Bj Schi` should bring up `Björn Schi` as well (space as a "and"-separator for better results)

As discussed @schiessle

### PR DESCRIPTION
**Pull request by dominikmebs**
Fixes #1963

@jancborchardt wrote the best way to start is to pick a issues with label "good first issue".
that´s why i just choosed this issue here from the list. :pick:  

- with this change you find user "Dominik Meyer" also when you search for "Dom Me" in the share dialog
- with this change you find user "Dominik Meyer" also when you enter his local share "uid@domain:port"

this is my first contribution so be friendly :smiley_cat: 

PS next to my code contribution is saw this code-comment:

> // sqlite doesn't like re-using a single named parameter here

i not understand this comment exactly, what is not possible with sqlite?
my change works in sqlite even tho i reuse the same column (uid), but i guess this comment means something else.

PPS i personally would love :smiling_face_with_three_hearts:  to find the user if you start with typing the lastname like "Meyer Dominik". but this is out of scope here i guess.

### CODE REVIEW

**Comment by kinimodmeyer** [comment]
"continuous-integration/drone/pr" failed but i don´t see the reason in https://drone.nextcloud.com/nextcloud/server/33829/14/5 ? can someone give me a hint or is it a pipeline issue? how is the process overall here now.

**Comment by nickvergessen** [comment]
It had a timeout, it's okay to ignore it.

File: lib/private/User/Database.php [hunk 1]
```diff
@@ -274,9 +275,20 @@ public function getDisplayNames($search = '', $limit = null, $offset = null) {
...
+
+		$serverAbsolutePath = \OC::$server->getURLGenerator()->getAbsoluteURL('/');
+		$serverFederationPath = rtrim(\OC\Share\Share::removeProtocolFromUrl($serverAbsolutePath), '/');
+
+		/*
+		 * the given search looks like a local federation.
+		 */
+		if (mb_strpos($search, '@'.$serverFederationPath) !== false) {
+			$query->orWhere($query->expr()->eq('uid', $query->createPositionalParameter(explode('@', $search)[0])));
+		}
```
Comment (nickvergessen) [inline lib/private/User/Database.php:289]: This should not be here from my pov.

File: lib/private/User/Database.php [hunk 2]
```diff
@@ -265,6 +265,7 @@ public function getDisplayNames($search = '', $limit = null, $offset = null) {
 
 		$query = $this->dbConn->getQueryBuilder();
 
+		$displayNameParameter = '%' . str_replace(' ', '% ', $this->dbConn->escapeLikeParameter($search)) . '%';
```
Comment (nickvergessen) [inline lib/private/User/Database.php:268]:
```suggestion
		$displayNameParameter = '%' . implode('% %', array_map([$this->dbConn, 'escapeLikeParameter'], explode(' ', $search)))) . '%';
```

**Review by nickvergessen — CHANGES_REQUESTED** [review]

File: lib/private/User/Database.php (hunk 1 above)
Comment (kinimodmeyer) [reply lib/private/User/Database.php:289]: thanks for your review @nickvergessen 
can you give me a little more information what you mean with this?
do you mean it should be in another class/method and somehow gets injected there?

Comment (nickvergessen) [reply lib/private/User/Database.php:289]: Well "federation a like" results are returned by other search plugins?

**Pull request closed without merging**